File: macos/Sources/AgentTimeline/Parsers/CodexParser.swift
import Foundation

/// ~/.codex/sessions/YYYY/MM/DD/rollout-<ts>-<uuid>.jsonl
/// Lines are {timestamp, type, payload}; user prompts are
/// type=="event_msg" && payload.type=="user_message".
struct CodexParser: AgentSessionParser {
    let agent = AgentKind.codex
    let root = ParserSupport.home("~/.codex/sessions")

    func watchRoots() -> [URL] { [root] }

    func makeContext(for url: URL) -> ParsedFileContext? {
        guard url.pathExtension == "jsonl",
              url.lastPathComponent.hasPrefix("rollout-"),
              url.path.hasPrefix(root.path) else { return nil }
        // rollout-2026-04-23T11-21-47-<uuid>.jsonl → trailing uuid as fallback id
        let stem = url.deletingPathExtension().lastPathComponent
        let sessionId = stem.split(separator: "-").suffix(5).joined(separator: "-")
        var context = ParsedFileContext(url: url, agent: .codex, sessionId: sessionId, project: "codex", cwd: nil)
        // session_meta is the first line; when resuming past a persisted offset
        // (app restart mid-session) we'd otherwise lose cwd/project and the
        // summarizer-scratch exclusion. Re-seed from the file head.
        // ⚠ 不能只读固定一小块：codex 的 session_meta 首行常常很大（本机 260 个
        // rollout 里 169 个 >16KB，最大 44KB），读少了找不到换行就整条放弃，
        // 重启续扫后项目名会退化成 "codex"（实测 379/1874 条命令受影响）。
        // 分块读到第一个换行为止，上限兜住病态的单行巨文件。
        if let handle = try? FileHandle(forReadingFrom: url) {
            var head = Data()
            let chunk = 64 * 1024
            let cap = 1024 * 1024
            while head.count < cap {
                guard let piece = try? handle.read(upToCount: chunk), !piece.isEmpty else { break }
                head.append(piece)
                if head.firstIndex(of: 0x0A) != nil { break }
            }
            if let nl = head.firstIndex(of: 0x0A),
               let line = String(data: head[head.startIndex..<nl], encoding: .utf8) {
                _ = parse(line: line, context: &context)
            }
            try? handle.close()
        }
        return context
    }

    /// `[$plugin:skill](本机 …/SKILL.md)` 开头的技能回显：留徽标文字、剥本机
    /// 绝对路径（跨机无效且泄漏用户名）。与 win CodexParser.SkillEchoRegex 同语义。
    private static let skillEchoRegex = try! NSRegularExpression(
        pattern: #"^\[(\$[^\]\n]+)\]\([^)\n]*SKILL\.md\)"#)

    static func convertSkillEcho(_ text: String) -> String {
        let range = NSRange(text.startIndex..., in: text)
        guard let m = skillEchoRegex.firstMatch(in: text, range: range),
              let badge = Range(m.range(at: 1), in: text),
              let whole = Range(m.range, in: text) else { return text }
        return (String(text[badge]) + text[whole.upperBound...])
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }

    /// A2（Windows 本机 168 万行语料实证）：codex 的 user_message 里混着编排器
    /// 注入块。`<task>…</task>` 是**给用户真实任务加的壳**（72 条）→ 去壳保留正文；
    /// 其余标签是自动化自发的（`<heartbeat>` 等）→ 整条跳过。修前 37 个节点的
    /// 标题字面就是 `<task>`。
    private static let taskWrapperRegex = try! NSRegularExpression(
        pattern: #"^<task>\s*([\s\S]*?)\s*</task>\s*$"#)

    private static let ignoredCodexBlocks = [
        "<user_instructions", "<environment_context", "<heartbeat",
        "<environments_instructions", "<apps_instructions", "<skills_instructions",
        "<plugins_instructions", "<collaboration_mode", "<multi_agent_mode",
        "<context_window", "<turn_aborted",
    ]

    /// - Returns: `nil` 表示整条跳过；否则是去壳后的正文。
    static func unwrapInjectedBlock(_ text: String) -> String? {
        let lower = text.lowercased()
        if ignoredCodexBlocks.contains(where: { lower.hasPrefix($0) }) { return nil }
        let range = NSRange(text.startIndex..., in: text)
        if let m = taskWrapperRegex.firstMatch(in: text, range: range),
           let inner = Range(m.range(at: 1), in: text) {
            let body = String(text[inner]).trimmingCharacters(in: .whitespacesAndNewlines)
            return body.isEmpty ? nil : body
        }
        return text
    }

    func parse(line: String, context: inout ParsedFileContext) -> [SessionEvent] {
        guard let obj = ParserSupport.json(line), let type = obj["type"] as? String else { return [] }
        let payload = obj["payload"] as? [String: Any] ?? [:]
        // 同上：session_meta / response_item 的时间戳也喂养回退基准（与 win 同口径）。
        let lineTimestamp = ParserSupport.timestamp(obj["timestamp"], carriedBy: &context)

        switch type {
        case "session_meta":
            // §4.2b B1：**只应用本文件第一条** session_meta。
            // 被 resume/fork 的 rollout 在文件中途还会写入**原会话**的 meta，逐条重设
            // 会让「实时扫」与「重启续扫」（后者只读第 0 行）判出不同 sessionId ——
            // sessionId 参与节点 id/唯一键，于是重扫插出重复行（本机库里 codex 已有
            // 38 组 / 41 行重复）；且结果行会被挂到**另一个 rollout 文件**的命令上。
            // 只认首条后：每个 rollout 自成会话，结果行只挂同文件的命令，语义更正确。
            guard !context.metaApplied else { return [] }
            context.metaApplied = true
            if let id = payload["id"] as? String { context.sessionId = id }
            if let cwd = payload["cwd"] as? String {
                context.cwd = cwd
                context.project = (cwd as NSString).lastPathComponent
                if cwd == AppSettings.summarizerScratchDir {
                    context.disabled = true
                }
            }
            return []

        case "event_msg":
            guard !context.disabled else { return [] }
            let payloadType = payload["type"] as? String
            guard let ts = lineTimestamp else { return [] }

            if payloadType == "user_message" {
                guard let raw = payload["message"] as? String else { return [] }
                // 与 win 一致：先 trim 再判定与落库（否则同一条命令两端正文差
                // 空白，连节点 id 都不同）。
                guard let unwrapped = Self.unwrapInjectedBlock(
                    raw.trimmingCharacters(in: .whitespacesAndNewlines)) else { return [] }
                let text = Self.convertSkillEcho(unwrapped)
                guard !ParserSupport.isIgnoredContent(text) else { return [] }
                let cmd = UserCommand(
                    agent: .codex,
                    project: context.project,
                    cwd: context.cwd,
                    sessionId: context.sessionId,
                    timestamp: ts,
                    text: text,
                    sourceFile: context.url.path)
                return [.userCommand(cmd)]
            }
            if payloadType == "task_complete" {
                guard let last = payload["last_agent_message"] as? String, !last.isEmpty else { return [] }
                return [.assistantText(agent: .codex, sessionId: context.sessionId, timestamp: ts, text: last)]
            }
            return []

        default:
            return []
        }
    }
}

File: windows/AgentTimeline/Core/Parsers/CodexParser.cs
using System.Text.Json;
using System.Text.RegularExpressions;

namespace AgentTimeline.Core.Parsers;

/// <summary>
/// Codex sessions — docs/SESSION-FORMATS.md §2.
///
/// Path:   %USERPROFILE%\.codex\sessions\YYYY\MM\DD\rollout-&lt;ts&gt;-&lt;uuid&gt;.jsonl
/// Format: each line {timestamp, type, payload}; type ∈ session_meta / event_msg /
///         response_item / turn_context / compacted.
///
///   - session_meta:  payload.id (session id), payload.cwd, payload.cli_version;
///   - user command:  type=="event_msg" &amp;&amp; payload.type=="user_message" → payload.message,
///     skipping environment injections starting with &lt;user_instructions&gt; or &lt;environment_context&gt;;
///   - task complete: payload.type=="task_complete" → payload.last_agent_message.
///
/// session_meta is the FIRST line of a rollout file. When tailing resumes mid-file after an
/// app restart the meta line is before our offset, so EnsureMeta re-reads just the first line.
///
/// `session_meta.payload.cwd` 等于摘要器 scratch 目录时**整文件禁用**（见 FileContext.Disabled）：
/// 那份 rollout 是我们自己 `codex exec` 摘要跑出来的，收进来就是自摄取回路。
/// </summary>
public sealed class CodexParser : IAgentSessionParser
{
    public AgentKind Agent => AgentKind.Codex;

    /// <summary>[$plugin:skill](…SKILL.md) 技能调用回显 → 只留 $plugin:skill 徽标文字。</summary>
    private static readonly Regex SkillEchoRegex = new(
        @"^\[(\$[^\]\n]+)\]\([^)\n]*SKILL\.md\)",
        RegexOptions.Compiled | RegexOptions.CultureInvariant | RegexOptions.IgnoreCase);

    /// <summary>&lt;task&gt;…&lt;/task&gt; 编排器包装 → 取内文（芯是用户真实任务）。</summary>
    private static readonly Regex TaskWrapperRegex = new(
        @"^<task>\s*(.*?)\s*</task>\s*$",
        RegexOptions.Compiled | RegexOptions.CultureInvariant | RegexOptions.Singleline);

    /// <summary>
    /// codex 侧非人类输入块（裸标签名，容忍带属性）。语料实证：`&lt;heartbeat&gt;` 是
    /// 自动化协调循环自发的轮次（含 automation_id/current_time_iso/instructions），
    /// 不是用户命令；`&lt;user_instructions&gt;`/`&lt;environment_context&gt;` 是环境注入。
    /// </summary>
    private static readonly string[] IgnoredBlocks =
    {
        "<user_instructions", "<environment_context", "<heartbeat",
        "<environments_instructions", "<apps_instructions", "<skills_instructions",
        "<plugins_instructions", "<collaboration_mode", "<multi_agent_mode",
        "<context_window", "<turn_aborted",
    };

    private static bool IsIgnoredCodexBlock(string text)
    {
        foreach (var tag in IgnoredBlocks)
        {
            if (text.StartsWith(tag, StringComparison.OrdinalIgnoreCase)) return true;
        }
        return false;
    }

    private sealed class FileContext
    {
        public string? SessionId;
        public string? Cwd;
        public bool MetaChecked;

        /// <summary>已应用过本文件的**第一条** session_meta（§4.2b B1）。</summary>
        public bool MetaApplied;

        /// <summary>本文件里最后一个**成功解析**的时间戳（W-e 回退基准）。</summary>
        public DateTimeOffset? LastTimestamp;

        /// <summary>
        /// 整文件忽略：这份 rollout 是**我们自己的摘要器**跑 `codex exec` 产生的。
        ///
        /// 摘要引擎解析到 codex 时，CliSummarizer 以 cwd=%LOCALAPPDATA%\AgentTimeline\summarizer
        /// 起进程，codex 把每条摘要 prompt 原样写成 `user_message` 落进
        /// `~\.codex\sessions\YYYY\MM\DD\rollout-*.jsonl`——路径里不含 "AgentTimeline"/"summarizer"，
        /// SessionWatcher 的路径级排除完全够不着。不认这个 cwd，时间线就会把自己发出的
        /// 每条摘要 prompt 当成用户命令收进来（自摄取回路）。mac 侧同判定。
        /// </summary>
        public bool Disabled;
    }

    private readonly Dictionary<string, FileContext> _contexts = new();

    public bool CanHandle(string path) =>
        path.EndsWith(".jsonl", StringComparison.OrdinalIgnoreCase) &&
        Path.GetFileName(path).StartsWith("rollout-", StringComparison.OrdinalIgnoreCase) &&
        path.Contains(Path.Combine(".codex", "sessions"), StringComparison.OrdinalIgnoreCase);

    public IReadOnlyList<SessionEvent> ParseLines(string path, IReadOnlyList<RawLine> lines)
    {
        var events = new List<SessionEvent>();
        if (lines.Count == 0) return events;

        if (!_contexts.TryGetValue(path, out var ctx))
        {
            ctx = new FileContext();
            _contexts[path] = ctx;
        }
        // Resuming mid-file without meta? Recover session_meta from line 0.
        if (ctx.SessionId is null && !ctx.MetaChecked && lines[0].ByteOffset > 0)
        {
            EnsureMeta(path, ctx);
        }
        if (ctx.Disabled) return events;

        foreach (var line in lines)
        {
            if (string.IsNullOrWhiteSpace(line.Text)) continue;
            try
            {
                using var doc = JsonDocument.Parse(line.Text);
                var root = doc.RootElement;
                if (root.ValueKind != JsonValueKind.Object) continue;

                var type = GetString(root, "type");
                // 时间戳（W-e，双端共同规则）：形态宽松解析；解不出就沿用本文件最后一个
                // 成功解析的时间戳；本文件还没有过任何时间戳则这一行不产出事件。
                // 绝不回退 UtcNow——那会让节点跳到时间线顶部，且 ts 参与
                // UNIQUE(agent,session_id,ts,command_hash)，重扫必产生重复行。
                var parsed = ParserUtil.TryParseIsoTimestamp(GetString(root, "timestamp"));
                if (parsed is not null) ctx.LastTimestamp = parsed;
                var timestamp = ctx.LastTimestamp;
                if (!root.TryGetProperty("payload", out var payload) ||
                    payload.ValueKind != JsonValueKind.Object)
                {
                    continue;
                }

                switch (type)
                {
                    case "session_meta":
                        ApplyMeta(ctx, payload);
                        if (ctx.Disabled) return events; // 摘要器自己的 rollout：整文件零事件
                        break;

                    case "event_msg":
                        if (ctx.Disabled || timestamp is not { } ts) break;
                        var payloadType = GetString(payload, "type");
                        if (payloadType == "user_message")
                        {
                            var message = GetString(payload, "message");
                            if (string.IsNullOrWhiteSpace(message)) break;
                            var text = message.Trim();
                            // 环境注入 / 自动化循环发起的轮次，都不是人打的字。
                            // ⚠ 标签一律**裸标签名匹配**（不含 '>'）——harness 会给注入块
                            // 带属性，带 '>' 的前缀匹配不上（与 Claude 侧同一教训）。
                            if (IsIgnoredCodexBlock(text)) break;
                            // 编排器把用户真实任务包在 <task>…</task> 里下发（本机语料 72 条）：
                            // 壳不是人写的，芯是——去壳留正文，否则时间线上 37 个节点的标题
                            // 字面就是 "<task>"（实机审计发现）。
                            if (TaskWrapperRegex.Match(text) is { Success: true } wrapped)
                            {
                                text = wrapped.Groups[1].Value.Trim();
                                if (text.Length == 0) break;
                            }
                            // 插件技能调用回显 [$plugin:skill](本地 SKILL.md 绝对路径) 开头
                            // (语料 17/70 条):保留命令徽标文字,剥掉本机路径(跨机无效
                            // 且泄漏用户名)。见 docs/TEXT-NORMALIZATION.md。
                            text = SkillEchoRegex.Replace(text, "$1", 1).Trim();

                            events.Add(new UserCommand(
                                Agent: AgentKind.Codex,
                                Project: ParserUtil.ProjectNameFromCwd(ctx.Cwd, fallback: "codex"),
                                SessionId: SessionIdFor(path, ctx),
                                Timestamp: ts,
                                Text: text,
                                SourceFile: path,
                                SourceOffset: line.ByteOffset));
                        }
                        else if (payloadType == "task_complete")
                        {
                            var last = GetString(payload, "last_agent_message");
                            if (!string.IsNullOrWhiteSpace(last))
                            {
                                events.Add(new TaskComplete(
                                    Agent: AgentKind.Codex,
                                    SessionId: SessionIdFor(path, ctx),
                                    Timestamp: ts,
                                    ResultLine: ParserUtil.ResultExcerpt(last),
                                    FullText: last)); // untruncated — mined for codenames
                            }
                        }
                        break;

                    // response_item / turn_context / compacted → agent internals, ignored.
                }
            }
            catch (JsonException)
            {
                // Skip malformed line.
            }
        }
        return events;
    }

    private static string SessionIdFor(string path, FileContext ctx) =>
        ctx.SessionId ?? Path.GetFileNameWithoutExtension(path);

    /// <summary>
    /// session_meta 落到上下文——首行直读（EnsureMeta）与增量流式两条路径必须**同一份逻辑**，
    /// 否则重启续扫时摘要器排除会漏掉（自摄取正是在重启后最容易复现）。
    /// </summary>
    private static void ApplyMeta(FileContext ctx, JsonElement payload)
    {
        // §4.2b B1：只应用**本文件第一条** session_meta。被 resume/fork 的 rollout
        // 在文件中途还会写入**原会话**的 meta，逐条重设会让「实时扫」与「重启续扫」
        // （后者只读第 0 行）判出不同 sessionId —— sessionId 参与唯一键，重扫因此
        // 插出重复行；且结果行会被挂到另一个 rollout 文件的命令上。
        ctx.MetaChecked = true;
        if (ctx.MetaApplied) return;
        ctx.MetaApplied = true;
        ctx.SessionId = GetString(payload, "id");
        ctx.Cwd = GetString(payload, "cwd");
        if (AppPaths.IsSummarizerWorkDir(ctx.Cwd)) ctx.Disabled = true;
    }

    private void EnsureMeta(string path, FileContext ctx)
    {
        ctx.MetaChecked = true;
        try
        {
            using var stream = new FileStream(path, FileMode.Open, FileAccess.Read,
                FileShare.ReadWrite | FileShare.Delete);
            using var reader = new StreamReader(stream);
            var first = reader.ReadLine();
            if (first is null) return;
            using var doc = JsonDocument.Parse(first);
            var root = doc.RootElement;
            if (GetString(root, "type") == "session_meta" &&
                root.TryGetProperty("payload", out var payload) &&
                payload.ValueKind == JsonValueKind.Object)
            {
                ApplyMeta(ctx, payload);
            }
        }
        catch (Exception ex)
        {
            Log.Warn($"CodexParser: failed to recover session_meta from {path}: {ex.Message}");
        }
    }

    private static string? GetString(JsonElement el, string name) =>
        el.TryGetProperty(name, out var p) && p.ValueKind == JsonValueKind.String ? p.GetString() : null;
}
